core/src/test/java/org/apache/calcite/util/SourceTest.java

```diff
@@ -67,12 +70,27 @@ private static String getRootPrefix() {
     }
   }
 
+  @Test public void testFileToUrl() throws URISyntaxException {
+    final String path = "abc.txt";
+    Source fileSource = file(null, path);
+    URL url = fileSource.url();
+
+    assertFalse(url == null);
+
+    String urlPath = new File(url.toURI().getPath()).getAbsolutePath();
+    assertEquals(absolutePath(path), urlPath);
+  }
+
+  private String absolutePath(String relativePath) {
+    return new File(relativePath).getAbsolutePath();
+  }
+
   @Test public void testAppendWithSpaces() {
     String fooRelative = "fo o+";
     String fooAbsolute = ROOT_PREFIX + "fo o+";
     String barRelative = "b ar+";
     String barAbsolute = ROOT_PREFIX + "b ar+";
-    assertAppend(file(null, fooRelative), file(null, barRelative), "fo o+/b ar+");
+    assertAppend(file(null, fooRelative), file(null, barRelative), absolutePath("fo o+/b ar+"));
```

Comment:
This looks like a behavior change. Can you please clarify what is produced in the output?

Can you please clarify why the append of two relative paths becomes an absolute? It does look weird.

Reply:
Sure. When the file source is constructed solely by a file object, a url is converted from the file object. Meantime, the path of the url is changed to an absolute path. When verifying results, it depends on the toString method of the file source, which uses the url path (as you have observed in a previous comment).

This is why the path has changed to an absolute one. 

I have made this explicit by adding a comment in the code.

Reply:
> Meantime, the path of the url is changed to an absolute path.

Does that mean the change prevents the use of `FileSource` with relative paths?
What's the reason to always convert the path to absolute format?

Reply:
> Does that mean the change prevents the use of FileSource with relative paths?

Maybe not. We can still provide a relative path to the constructor of FileSource. The constructor will convert it to an absolute path and use it to construct a url. FileSource still works appropriately.

> What's the reason to always convert the path to absolute format?

This is determined by the behavior of the standard Java API. But I think the fundamental reason is that, according to the definition (https://en.wikipedia.org/wiki/URL), a URL is always based on an absolute path, as it needs to identify a resource without the aid of other information (e.g. working directory).

Reply:
Just to recap:  the current code supports relative paths.
Now you came and suggest a fix that (silently) drops that support.

I do not like features being dropped for no reason :-/

Reply:
@vlsi 
Thanks a lot for the feedback. I agree with you that we should not drop a feature silently. 
But sorry I do not fully understand your point. 

IMO, we do not drop the support for relative path, as we still can construct a file source with a relative file path, and the source works appropriately. What we change is the internal stuff of the class. So I am not sure what is the feature referred to here.

Maybe the feature refers to the output format of the toString method? I think it is sth. that can be overcome, but I am not sure if it is necessary. It seems the method is only used in the test case, and I do not find any evidence suggesting it is part of the class specification.  

Would you please give more information/hints?